prebackup/prebackup.go

```diff
@@ -190,43 +227,76 @@ func (p *PreBackup) getClientset() (*kubernetes.Clientset, error) {
+			endWatch, err := p.handleEvent(event)
+			if err != nil {
+				return err
 			}
-
-			if hasAvailableReplica(deployment) {
+			if endWatch {
 				return nil
 			}
+		case <-p.CTX.Done():
+			p.Log.Error(p.CTX.Err(), "unexpected end during deployment watch ", "deployment", fmt.Sprintf("%s/%s", p.Obj.GetMetaObject().GetNamespace(), p.Obj.GetMetaObject().GetName()))
+			return p.CTX.Err()
+		}
+	}
+}
+
+func (p *PreBackup) handleEvent(event watch.Event) (bool, error) {
+	deployment, ok := event.Object.(*appsv1.Deployment)
+	if !ok {
+		p.Log.V(2).Info("unexpected event during deployment watch ", "event source", event.Object, "event type", event.Type)
+		return false, nil
+	}
 
-			p.Log.Info("waiting for command pod to get ready", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name))
+	p.Log.V(2).Info("new event during deployment watch ", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name), "event type", event.Type)
 
-		case watch.Error:
+	switch event.Type {
+	case watch.Modified:
+		ready, err := p.isDeploymentReady(deployment)
+		if err != nil || ready {
+			return true, err
+		}
 
-			last := p.getLastDeploymentCondition(deployment)
+		p.Log.Info("waiting for command pod to get ready", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name))
 
-			if last != nil {
-				return fmt.Errorf("there was an error while starting pre backup pod '%v/%v': %v", deployment.Namespace, deployment.Name, last.Message)
-			}
-			return fmt.Errorf("there was an unknown error while starting pre backup pod '%v/%v'", deployment.Namespace, deployment.Name)
-
-		case watch.Added:
-		case watch.Bookmark:
-		case watch.Deleted:
-			p.Log.V(1).Info("ignoring event", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name), "event type", event.Type)
-		default:
-			p.Log.Info("unexpected event during deployment watch ", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name), "event type", event.Type)
+	case watch.Error:
+		last := p.getLastDeploymentCondition(deployment)
+
+		var err error
+		if last != nil {
+			err = fmt.Errorf("there was an error while starting pre backup pod '%v/%v': %v", deployment.Namespace, deployment.Name, last.Message)
+		} else {
+			err = fmt.Errorf("there was an unknown error while starting pre backup pod '%v/%v'", deployment.Namespace, deployment.Name)
 		}
+		return true, err
+
+	case watch.Added:
+	case watch.Bookmark:
+	case watch.Deleted:
+		p.Log.V(1).Info("ignoring event", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name), "event type", event.Type)
+
+	default:
+		p.Log.Info("unexpected event during deployment watch ", "deployment", fmt.Sprintf("%s/%s", deployment.Namespace, deployment.Name), "event type", event.Type)
 	}
+	return false, nil
+}
 
-	return nil
+func (p *PreBackup) isDeploymentReady(deployment *appsv1.Deployment) (bool, error) {
+	last := p.getLastDeploymentCondition(deployment)
```

Comment:
We might want to check whether the deployment has a deletion date but still existed at the time when doing GET.

Reply:
Let's hope for now that this is not a use case we encounter often.